From 62f949d7dc3b6b7a7a977bdb63f9ac432099e3b6 Mon Sep 17 00:00:00 2001
From: Joerg Bornemann <joerg.bornemann@qt.io>
Date: Wed, 17 Sep 2025 17:16:16 +0200
Subject: [PATCH 151/553] CMake: Handle library names like libavformat.a in
 .prl files

We generated faulty .prl files for Qt modules that link against library
file names like "libavformat.a". The .prl file contained wrong LIBS
entries like "-llibavformat.a".

Now, we check whether a library entry looks like a library file name and
don't prepend the -l in that case.

Fixes: QTBUG-138596
Pick-to: 6.8
Change-Id: I44ce5de91e4575bb45d391db4f1d3d70efefcdfa
Reviewed-by: Alexandru Croitor <alexandru.croitor@qt.io>
(cherry picked from commit 121106ae19423997ced4846f1e30dfbe9800819a)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtFinishPrlFile.cmake      |  5 ++++-
 cmake/QtGenerateLibHelpers.cmake | 19 +++++++++++++++++++
 2 files changed, 23 insertions(+), 1 deletion(-)

diff --git a/cmake/QtFinishPrlFile.cmake b/cmake/QtFinishPrlFile.cmake
index 0cef22617cb..3c7caebaa15 100644
--- a/cmake/QtFinishPrlFile.cmake
+++ b/cmake/QtFinishPrlFile.cmake
@@ -78,7 +78,10 @@ foreach(line ${lines})
                 # Not absolute path, most likely a library name or a linker flag.
                 # If linker flag (like -framework, -lfoo, -pthread, keep it as-is).
                 if(NOT lib MATCHES "^-")
-                    string(PREPEND lib "-l")
+                    qt_is_library_file(is_library_file "${lib}")
+                    if(NOT is_library_file)
+                        string(PREPEND lib "-l")
+                    endif()
                 endif()
                 list(APPEND adjusted_libs "${lib}")
             endif()
diff --git a/cmake/QtGenerateLibHelpers.cmake b/cmake/QtGenerateLibHelpers.cmake
index be998d0e6bd..dbce9973658 100644
--- a/cmake/QtGenerateLibHelpers.cmake
+++ b/cmake/QtGenerateLibHelpers.cmake
@@ -134,3 +134,22 @@ function(qt_internal_path_is_relative_to_qt_lib_path
     set(${out_var_is_relative} "${is_relative}" PARENT_SCOPE)
     set(${out_var_relative_path} "${relative_path_value}" PARENT_SCOPE)
 endfunction()
+
+# Checks if a filename looks like a library file (e.g., libfoo.a, /usr/lib/libbar.so.1.2.3).
+# Returns TRUE if it's a library file, FALSE otherwise.
+function(qt_is_library_file out_var file_path)
+    set(is_library FALSE)
+
+    qt_strip_library_version_suffix(file_path "${file_path}")
+    get_filename_component(ext "${file_path}" EXT)
+    if(ext)
+        foreach(libsuffix ${LIBRARY_SUFFIXES})
+            if(ext STREQUAL libsuffix)
+                set(is_library TRUE)
+                break()
+            endif()
+        endforeach()
+    endif()
+
+    set(${out_var} "${is_library}" PARENT_SCOPE)
+endfunction()
-- 
2.51.2

